---
title: "Introduction to Bulk RNAseq data analysis"
subtitle: Differential Expression of RNA-seq data - Part 1
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_document:
    toc: yes
  pdf_document:
    toc: yes
bibliography: ref.bib
---

```{r setup, echo=FALSE}
options(tibble.print_max = 4, tibble.print_min = 4, max.print=40, 
        tibble.max_extra_cols=2)
```

## Recap of pre-processing

Where are we in the workflow? What has been done so far, step by step

What did Ash do this morning? mention filtering of low count genes

DESEQ2
- deseq vs edgeR, slightly different normalisation and filtering methods
 [here](https://mikelove.wordpress.com/2016/09/28/deseq2-or-edger/) by Mike Love
 
- help, forums, workflow page

## Start Coding ~ 15mins

have this copy pasted in at start

```{r}
# Read the sample information into a data frame
#sampleinfo <- read_tsv("results/SampleInfo_Corrected.txt")
# Read the data into R
#seqdata <- read_tsv("data/GSE60450_Lactation.featureCounts", comment = "#")
# Transform the data to matrix of counts
#countdata <- as.data.frame(seqdata) %>% 
    #column_to_rownames("Geneid") %>% # turn the geneid column into rownames
    #rename_all(str_remove, ".bam") %>% # remove the ".bam" from the column names
    #select(sampleinfo$Sample) %>% # keep sample columns using sampleinfo
    #as.matrix()
# filter the data to remove genes with few counts
#keep <- rowSums(countdata) > 5
#countdata <- countdata[keep,]
```

Now that we are happy that the quality of the data looks good, proceed

- clean environment

- restart session

- make new script

## Load the data

This morning txi object saved and sample metadata table

First load the packages we need.

```{r message = FALSE}
library(DESeq2)
library(tidyverse)
```

Now load the data from the earlier session.

```{r loadData}
txi <- readRDS("RObjects/txi.rds")
sampleinfo <- read_tsv("data/samplesheet_corrected.tsv", col_types="cccc")
```

Explain and view everything that has been loaded

Briefly sample info tab: status, celltype, 2 reps

It is important to be sure that the order of the samples in rows in the sample 
meta data table matches the order of the columns in the data matrix - `DESeq2`
will **not** check this. If the order does not match you will not be running the
analyses that you think you are.

```{r checkSampleNames}
all(colnames(txi$counts)==sampleinfo$SampleName)
```

### Create a DESeqDataSet object with the raw data

##### * Count data
##### * Meta data 
##### * Model

# The model formula and design matrices

Hugo talked about this

what we think is affecting expression, so for us that would be the status and timepoint

BUT for simplicity to learn workflow, simple model just status and ignore timepoint, we will go into other models next week.


## Create a DESeqDataSet object with the raw data

### Creating the design model formula - 23 mins


```{r modelForumla}
simple.model <- as.formula(~ Status)
```

~ means something on left is related to something on the right

- we don't have to actually put anything on the left as we know its the expression data 

- right side has to be a column header


####  model matrix, deseq2 makes it internally but we can get one to see what it looks like

```{r modelMatrix}
model.matrix(simple.model, data = sampleinfo)
```

it as automatically selected on condition as the intercept/baseline (alphabetical)

explain each coeffient

infected is not the ideal intercept so lets change it to uninfected, have to set the factor levels manually


```{r setFactors}
sampleinfo <- mutate(sampleinfo, Status = fct_relevel(Status, "Uninfected"))
model.matrix(simple.model, data = sampleinfo)
```

# Build a DESeq2DataSet ~ 33 mins

own type of object which holds all of our data in one place so we don't have to put all the bits in everytime

we give it the three things it needs

```{r makeDDSObj}
# create the DESeqDataSet object
ddsObj.raw <- DESeqDataSetFromTximport(txi = txi,
                                       colData = sampleinfo,
                                       design = simple.model)
```


Gives a warning, don't worry

## SAVE SCRIPT

# Filter out the unexpressed genes

Just as we did this morning, we should filter out genes that uninformative.

```{r}
keep <- rowSums(counts(ddsObj.raw)) > 5
ddsObj.filt <- ddsObj.raw[keep,]
```

# Differential expression analysis with DESeq2

## The `DESeq2` work flow ~ 35 mins

The main `DESeq2` work flow is carried out in 3 steps:

### Size factors

allows us to normalise between samples and compensate for sequencing depth and composition bias

### Dispersion

for each gene we need mean and dispersion (which relates mean to variance)

### Applying GLM and running Wald test

test each gene between the sample groups

## START


### `estimateSizeFactors`

First, Calculate the "median ratio" normalisation size factors...

keep same object, keep raw one safe

```{r commonSizeFactors}
ddsObj <- estimateSizeFactors(ddsObj.filt)
```

#### Let's have a look at what that did

`DESeq2` has calculated a normalizsation factor for each gene for each sample.

```{r}
normalizationFactors(ddsObj.filt)
normalizationFactors(ddsObj)
```

its added extra size factors column that it can use to normalise between samples

We can use `plotMA` from `limma` to look at the data in an MA plot

MA plot is very common way to visualise data

```{r}
logcounts <- log2(counts(ddsObj, normalized=FALSE)  + 1)

limma::plotMA(logcounts, array = 5, ylim =c(-5, 5))
abline(h=0, col="red")
```

this plots first column 

each point is gene

x axis is average expression

y axis is log ratio of expression of this sample vs. all other samples, so how different is this sample compared to other samples

A MAJOR ASSUMPTION with DEA is that in most conditions, most genes aren't changing across samples. There are some edge cases where this might not be true an if that is the case you might not be able to use the size factors to normalise.

expectation: The plot should be centred on the zero line

RAW: This is centered above the zero line (size factor is above 1 in table)


```{r}
logNormalizedCounts <- log2(counts(ddsObj, normalized=TRUE)  + 1)

limma::plotMA(logNormalizedCounts, array=5, ylim =c(-5, 5))
abline(h=0, col="red")
```

NEW: centred on zero line

## ... next estimate dispersion ... ~ 47 mins

Next we need to estimate the dispersion parameters for each gene.

```{r genewiseDispersion}
ddsObj <- estimateDispersions(ddsObj)
```

- gene wise , estimates for each gene but as we don't have that many samples the estimate isn't that great

- mean, by fitting the gene wise, gives something very conservative

SHOW DISPERSION PLOT

```{r plotDisp}
plotDispEsts(ddsObj)
```

black dots are gene wise estimation
red line is the fitting

The it does some sort of Bayesian statistics the shrink the gene wise estimates towards the fitted line to get a sort of inbetween that closer to the real dispersion, that is blue dots

### ...Finally, apply Negative Binomial GLM fitting and calculate Wald statistics.

```{r applyGLM}
ddsObj <- nbinomWaldTest(ddsObj)
```

## The `DESeq` command

In practice the 3 steps above can be performed in a single step using the 
`DESeq` wrapper function. Performing the three steps separately is useful if you
wish to alter the default parameters of one or more steps, otherwise the `DESeq`
function is fine.

```{r theShortVersion}
ddsObj <- DESeq(ddsObj.filt)
```

# SAVE SCRIPT

## Generate a results table

We can generate a table of differential expression results from the DDS object
using the `results` function of DESeq2.

```{r resultsTable}
results.simple <- results(ddsObj, alpha=0.05)
results.simple
```

Makes a DESeq results object.

This is table Dom was showing earlier at the start and end of his talk.

header - lactate over virgin 

+ve FC means expression in lactate mice is higher than in virgin mice and vis versa

SE of FC

Wald statistic

pvalue - knowing th distribution of the wald statistic how likely is it that null hypothesis is true

MTC  - Hugo explained, we are doing lots of tests and each additional test adds error. The default in DESeq2 is to use the Benjamini-Hochberg method which gives us FDR.

As you can see this is very simple to run, you can put anything in and get numbers out, the point of Dom's lecture earlier is almost to know what you don't know and when you might want to find your friendly local bioinformatician or statistian.


### Exercise 1

> Now we have made our results table using our simple model, let have a look at which
> genes are changing and how many pass our 0.05 threshold. Why might this not be
> straigtforward?
>
>

```{r }
sum(results.simple$padj < 0.05)
```

```{r }
sum(is.na(results.simple$padj))
```
>
> a) how many genes are significantly (with an FDR < 0.05) up-regulated?
>
>

```{r }
sum(results.simple$padj < 0.05 & results.simple$log2FoldChange > 0, na.rm = TRUE)
```
>
> b) how many genes are significantly (with an FDR < 0.05) down-regulated?
>
>

```{r }
sum(results.simple$padj < 0.05 & results.simple$log2FoldChange < 0, na.rm = TRUE)
```

### Independent filtering

You will notice that some of the adjusted p-values (`padj`) are NA. 

Deseq does independent filtering, the genes with `NA` are the ones `DESeq2` has filtered out.

Best to read inthe deseq manual but briefly: runs all this and looks at your alpha and sees how many pass the threshold on FDR, then it goes back and looks at the baseMean and finds the gene with the lowest expression and excludes it. then it re does the MTC. Slowly it drops out genes until it gets to a point where most genes pass. 

This might sound like cherrypicking but its ok because there is no coorelation between baseMean and pvalue.

Again best to read in manual but thats why there are NA's.
