---
title: "Introduction to Bulk RNAseq data analysis"
subtitle: "Gene Set Testing for RNA-seq"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_document:
    toc: yes
    toc_float: true
  pdf_document:
    toc: yes
bibliography: ref.bib
always_allow_html: true
---

```{r setup, echo=FALSE, cache=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 4, fig.height = 3, cache=TRUE)
```

The list of differentially expressed genes is sometimes so long that its 
interpretation becomes cumbersome and time consuming. It may also be very
short while some genes have low p-value yet higher than the given threshold.

A common downstream procedure to combine information across genes is gene set
testing. It aims at finding pathways or gene networks the differentially
expressed genes play a role in.

Various ways exist to test for enrichment of biological pathways. We will look
into over representation and gene set enrichment analyses.

A gene set comprises genes that share a biological function, chromosomal
location, or any other relevant criterion.

To save time and effort there are a number of packages that make applying these
tests to a large number of gene sets simpler, and which will import gene lists 
for testing from various sources.

Today we will use 
[`clusterProfiler`](https://yulab-smu.github.io/clusterProfiler-book/index.html).

# Over-representation

## Method

This method tests whether genes in a pathway are present in a subset of our data
in a higher number than expected by chance (explanations derived from the 
[clusterProfiler manual](https://yulab-smu.github.io/clusterProfiler-book/index.html)).

Genes in the experiment are split in two ways:

* annotated to the pathway or not
* differentially expressed or not

We can then create a contingency table with:

* rows: genes in pathway or not
* columns: genes differentially expressed or not

And test for independence of the two variables with the Fisher exact test.

## `clusterProfiler`

`clusterprofiler` [@Yu2012] supports direct online access of the current KEGG
database (KEGG: Kyoto Encyclopedia of Genes and Genomes), rather than relying on
R annotation packages.
It also provides some nice visualisation options.

We first search the resource for mouse data:

```{r loadClusterProfiler, message=FALSE}
library(tidyverse)
library(clusterProfiler)

search_kegg_organism('mouse', by='common_name')
```

We will use the 'mmu' 'kegg_code'.

## KEGG enrichment analysis

The input for the KEGG enrichment analysis is the list of gene IDs of significant genes. 

We now load the R object keeping the outcome of the differential expression 
analysis for the d11 contrast.

```{r loadData}
shrink.d11 <- readRDS("RObjects/Shrunk_Results.d11.rds")
```

We will only use genes that have:

* an adjusted p-value (FDR, False Discovery Rate) of less than 0.05
* and an absolute fold change greater than 2.

We need to remember to eliminate genes with missing values in the FDR as a result
of the independent filtering by DESeq2.

For this tool we need to use Entrez IDs, so we will also need to eliminate genes
with a missing Entrez ID (NA values in the 'Entrez' column).

```{r enrichKEGGprint}
sigGenes <- shrink.d11 %>% 
    drop_na(Entrez, padj) %>% 
    filter(padj < 0.05 & abs(log2FoldChange) > 1) %>% 
    pull(Entrez)

keggRes <- enrichKEGG(gene = sigGenes, organism = 'mmu')
as_tibble(keggRes)
```

### Visualise a pathway in a browser

`clusterProfiler` has a function `browseKegg` to view the KEGG pathway in a browser,
highlighting the genes we selected as differentially expressed.

We will show one of the top hits: pathway 'mmu04612' for 'Antigen processing and presentation'.

```{r browseKegg}
browseKEGG(keggRes, 'mmu04612')
```

### Visualise a pathway as a file

The package `pathview` [@Luo2013] can be used to generate figures of KEGG 
pathways. 

One advantage over the `clusterProfiler` browser method `browseKEGG` is that genes can be
coloured according to fold change levels in our data. To do this we need
to pass `pathview` a named vector of fold change values (one could in fact
colour by any numeric vector, e.g. p-value).

The package plots the KEGG pathway to a `png` file in the working directory.

```{r pathview, message=F}
library(pathview)
logFC <- shrink.d11$log2FoldChange
names(logFC) <- shrink.d11$Entrez
pathview(gene.data = logFC, 
         pathway.id = "mmu04612", 
         species = "mmu", 
         limit = list(gene=20, cpd=1))
```

*mmu04612.pathview.png*:

![mmu04612 - Antigen processing and presentation](images/mmu04612_pathview.png)

> ### Exercise 1 {.challenge}
>
> 1. Use `pathview` to export a figure for "mmu04659" or "mmu04658", but this 
> time only use genes that are statistically significant at padj < 0.01
>

# GO term enrichment analysis

`clusterProfiler` can also perform over-representation analysis on GO terms 
using the command `enrichGO`. For this analysis we will use Ensembl gene IDs 
instead of Entrez IDs and in order to do this we need to load another package 
which contains the mouse database called `org.Mm.eg.db`. 

To run the GO enrichment analysis, this time we also need a couple of extra 
things. Firstly, we should provide a list of the 'universe' of all the genes in 
our DE analysis not just the ones we have selected as significant. 

Gene Ontology terms are divided into 3 categories. 
- Metabolic Functions
- Biological Processes
- Cellular Components

For this analysis we will narrow our search terms in the 'Biological Processes'
Ontology so we can add the parameter "BP" with the 'ont' argument (the default 
is Molecular Functions). 

```{r GO analysis, message=FALSE, warning = FALSE}

library(org.Mm.eg.db)

sigGenes_GO <-  shrink.d11 %>% 
    drop_na(padj) %>% 
    filter(padj < 0.01 & abs(log2FoldChange) > 2) %>% 
    pull(GeneID)

universe <- shrink.d11$GeneID

ego <- enrichGO(gene          = sigGenes_GO, 
                universe      = universe,
                OrgDb         = org.Mm.eg.db,
                keyType       = "ENSEMBL",
                ont           = "BP",
                pvalueCutoff  = 0.01,
                readable      = TRUE)
```

We can use the `barplot` function to visualise the results. Count is the number 
of differentially expressed in each gene ontology term.

```{r barplot, fig.height=10, fig.width=8}
barplot(ego, showCategory=20)
```

or perhaps the `dotplot` version is more informative. Gene ratio is Count 
divided by the number of genes in that GO term.

```{r dotplot, fig.height=7, fig.width=8}
dotplot(ego, font.size = 14)
```

Another visualisation that can be nice to try is the `emapplot` which shows the overlap between genes in the different GO terms.

```{r emap, fig.height=8, fig.width=8}
library(enrichplot)
ego_pt <- pairwise_termsim(ego)
emapplot(ego_pt, cex.params = list(category_label = 0.8))
```


# GSEA analysis

Gene Set Enrichment Analysis (GSEA) identifies gene sets that are enriched in the dataset between samples [@Subramanian15545].

The software is distributed by the 
[Broad Institute](http://software.broadinstitute.org/gsea/index.jsp)
and is freely available for use by academic and non-profit organisations.
The Broad also provide a number of very well curated gene sets
for testing against your data - the [Molecular Signatures Database (MSigDB)](http://software.broadinstitute.org/gsea/msigdb/index.jsp). 

These gene lists are made availalble for R in the Bioconductor package `msigdb`
and the available dataset can be explored via `ExperimentHub`.

First, we need to locate the correct database and download the data.

```{r}
library(msigdb)
library(ExperimentHub)

eh = ExperimentHub()
query(eh , c('msigdb', 'mm', '2023'))
```

The most recent available release of MSigDb is "msigdb.v2023.1", so we'll
download this one. We have the option to use Entrez IDs or gene symbols. As we
already have Entrez IDs in our annotation, we'll use these. 

```{r}
msigdb.mm <- getMsigdb(org = 'mm', id = 'EZID', version = '2023.1')

msigdb.mm

listCollections(msigdb.mm)
```




## Method

The analysis is performed by:

1. ranking all genes in the data set  
2. identifying in the ranked data set the rank positions of all members of the 
gene set 
3. calculating an enrichment score (ES) that represents the difference between
the observed rankings and that which would be expected assuming a random rank
distribution.
 
The article describing the original software is available 
[here](http://www.pnas.org/content/102/43/15545.long),
while this 
[commentary on GSEA](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1266131/) 
provides a shorter description.

![](../images/gseaArticleFig1.png)

We will use `clusterProfiler`'s [`GSEA`](http://yulab-smu.top/clusterProfiler-book/chapter2.html#gene-set-enrichment-analysis) 
package [@Yu2012] that implements the same algorithm in R. 

## Rank genes

We need to provide `GSEA` with a vector containing values for a given gene
mtric, e.g. log(fold change), sorted in decreasing order.

To start with we will simply use a rank the genes based on their fold change.

We must exclude genes with no Entrez ID.

Also, we should use the shrunk LFC values.

```{r preparedata}
rankedGenes <- shrink.d11 %>%
  drop_na(GeneID, padj, log2FoldChange) %>%
  mutate(rank = log2FoldChange) %>%
  filter(!is.na(Entrez)) %>%
  arrange(desc(rank)) %>%
  pull(rank, Entrez)
head(rankedGenes)
```

## Load pathways

For `clusterProfiler` we need the genes and genesets to be in the form of is a `tibble`
with information on each {gene set; gene} pair in the rows.

```{r loadPathways_msigdb}
hallmarks = subsetCollection(msigdb.mm, 'h')
msigdb_ids = geneIds(hallmarks)

term2gene <- enframe(msigdb_ids, name = "gs_name", value = "entrez") %>%
  unnest(entrez)

head(term2gene)
```

## Conduct analysis

Arguments passed to `GSEA` include:

* ranked genes
* pathways
* gene set minimum size
* gene set maximum size

```{r runGsea, warning=FALSE}
gseaRes <- GSEA(rankedGenes,
                TERM2GENE = term2gene,
                pvalueCutoff = 1.00, 
                minGSSize = 15,
                maxGSSize = 500)
```

Let's look at the top 10 results.

```{r top10GseaPrint, eval=FALSE}
as_tibble(gseaRes) %>% 
  arrange(desc(abs(NES))) %>% 
  top_n(10, wt=-p.adjust) %>% 
  dplyr::select(-core_enrichment) %>%
  mutate(across(c("enrichmentScore", "NES"), ~round(.x, digits=3))) %>% 
  mutate(across(c("pvalue", "p.adjust", "qvalue"), scales::scientific))
```

```{r top10GseaPrintactual, echo=FALSE, warning=FALSE}
as_tibble(gseaRes) %>% 
  arrange(desc(abs(NES))) %>% 
  top_n(10, wt=-p.adjust) %>% 
  dplyr::select(-core_enrichment) %>%
  mutate(across(c("enrichmentScore", "NES"), ~round(.x, digits=3))) %>% 
  mutate(across(c("pvalue", "p.adjust", "qvalue"), scales::scientific)) %>% 
  DT::datatable(option=list(dom='t'))
```

## Enrichment score plot

The enrichment score plot displays along the x-axis that represents the decreasing gene rank:

* genes involved in the pathway under scrutiny: one black tick per gene in the
pathway (no tick for genes not in the pathway) 
* the enrichment score: the green curve shows the difference between the observed
rankings and that which would be expected assuming a random rank distribution.

```{r gseaEnrichmentPlot_both, fig.height=8, fig.width=10}
gseaplot(gseaRes, 
         geneSetID = "HALLMARK_INFLAMMATORY_RESPONSE", 
         title = "HALLMARK_INFLAMMATORY_RESPONSE")
```

Remember to check the [GSEA 
article](http://www.pnas.org/content/102/43/15545.full) for the complete 
explanation.

> ## Exercise 2 {.exercise}
>
> Another common way to rank the genes is to order by pvalue while sorting
> so that upregulated genes are at the start and downregulated at the end.
> You can do this combining the sign of the fold change and the pvalue. 
> 
> 1. Rank the genes by statistical significance - you will need to create
> a new ranking value using `-log10(pvalue) * sign(log2FoldChange)`.
> 2. Run GSEA using the new ranked genes and the H pathways.
> 3. Conduct the same analysis for the day 33 Infected vs Uninfected contrast.


---------------------------------------------------------------

# References
